### PR TITLE
Fix FittedBox BoxFit.scaleDown sizing

### LINKED ISSUE #42776: FittedBox.scaleDown can scale up the container used
Internal: b/142592677

Not sure if this is intentional, but it was unexpected from a usage standpoint (there are workarounds, so this is not critical).

With FittedBox widget removed:

image

With FittedBox widget:

image

Expected that using FittedBox.scaleDown would not increase the container size.

Full repro code:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: Scaffold(
        appBar: AppBar(title: Text('FittedBox test')),
        body: Container(
          color: Colors.red,
          child: Row(
            children: [
              Expanded(
                child: FittedBox(
                  fit: BoxFit.scaleDown,
                  child: Text('Leading'),
                ),
              ),
              Text('trailing'),
            ],
          ),
        ),
      ),
    );
  }
}
```

### PR DESCRIPTION
**Pull request by pingbird**
## Description

Currently FittedBox can be much bigger than the child when using BoxFit.scaleDown because it always attempts to preserve the child's aspect ratio (see attached issue for what that looks like).

This PR fixes that by adding a special case for BoxFit.scaleDown during layout, which will loosen the constrains passed to `constrainSizeAndAttemptToPreserveAspectRatio`, allowing FittedBox to match the size of its child better.

Because the current behavior of this corner case is awkward and not covered under existing tests, I don't think the change will likely affect customers.

## Related Issues

Fixes #42776

## Tests

I added the following tests:

1. BoxFit.scaleDown matches size of child
2. Switching to and from BoxFit.scaleDown causes relayout

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

**Comment by PixelToast** [comment]
Here is a pad you can play around with: https://dartpad.dartlang.org/b5460015397df846823ebe88f0a3394e

The modified FittedBox is on the right

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 1]
```diff
@@ -2352,16 +2352,30 @@ class RenderFittedBox extends RenderProxyBox {
     markNeedsPaint();
   }
 
+  bool _fitAffectsLayout(BoxFit fit) {
+    switch(fit) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart:2356]: trivial nit: please be consistent about whether a space appears after `switch` (space preferred)

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 2]
```diff
@@ -2352,16 +2352,30 @@ class RenderFittedBox extends RenderProxyBox {
...
+    }
+  }
+
   /// How to inscribe the child into the space allocated during layout.
   BoxFit get fit => _fit;
   BoxFit _fit;
-  set fit(BoxFit value) {
-    assert(value != null);
-    if (_fit == value)
+  set fit(BoxFit newFit) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart:2367]: per our style guide, please call the argument "value"

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 3]
```diff
@@ -2352,16 +2352,30 @@ class RenderFittedBox extends RenderProxyBox {
...
+    assert(newFit != null);
+    final BoxFit lastFit = _fit;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart:2369]: it's trivial, but ideally this line would be after the early return

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 4]
```diff
@@ -2403,7 +2417,18 @@ class RenderFittedBox extends RenderProxyBox {
   void performLayout() {
     if (child != null) {
       child.layout(const BoxConstraints(), parentUsesSize: true);
-      size = constraints.constrainSizeAndAttemptToPreserveAspectRatio(child.size);
+
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart:2420]: trivial nit: no need for these blank lines

**Comment by Hixie** [comment]
This is a very clean PR with nice tests, thanks!

I'm always a bit dubious about special cases like this, but I admit that this seems like the right answer in practice. Thanks for making the dartpad, that was really helpful in trying to explore this patch.

LGTM modulo the comments above.

**Review by Hixie — APPROVED** [review]
Thanks!

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 5]
```diff
@@ -2352,16 +2352,30 @@ class RenderFittedBox extends RenderProxyBox {
     markNeedsPaint();
   }
 
+  bool _fitAffectsLayout(BoxFit fit) {
```
Comment (Calamity210) [inline packages/flutter/lib/src/rendering/proxy_box.dart:2355]: Small change but;
```diff
+ return fit == BoxFit.scaleDown;
-   switch (fit) {
-      case BoxFit.scaleDown:
-        return true;
-      default:
-        return false;
-    }
-  }
```

Comment (PixelToast) [reply packages/flutter/lib/src/rendering/proxy_box.dart:2355]: switch cases are generally preferred over comparison in the repo, but so is the default case tbh

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values

**Pull request merged**